plat/virtio_pci: Fix buffer length argument when calling `virtio_cread_bytes_many`

eduardvintila opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [`x86_64`, `aarch64`]
 - Platform(s): [kvm]
 - Application(s): [redis, sqlite, nginx]


### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->
This PR fixes an incorrect call to the `virtio_cread_bytes_many` function which is used to read a virtio PCI device's configuration. The function expects a `len` argument which is the total size of the buffer where the configuration bytes will be copied. However, the `vpci_legacy_pci_config_get` function invokes this routine by specifying the number of elements in the buffer, not its total size in bytes. This results in issues for example in the 9pfs filesystem initialization phase, where the filesystem tag (usuallly `fs0`) is incorrectly read and could result in filesystem setup failure. Problems could even arise past the configuration phase. More detalis can be found here in Issue #518 made by @mariasfiraiala.

Signed-off-by: Eduard Vintilă <eduard.vintila47@gmail.com>

eduardvintila committed:
plat/virtio_pci: Fix buffer length argument in function call

In the `vpci_legacy_pci_config_get` function, `len` is equal to the
number of elements in the buffer, where each element is of size
`type_len` bytes.

In this routine, the `_virtio_cread_bytes` is called using these
two parameters if an entity has less then 4 bytes. However, when calling
`virtio_cread_bytes_many`, the function has only one argument for
specifying the buffer length, and it should be used to pass the total
size of the buffer in bytes (i.e. len * type_len), not the
number of elements.

Signed-off-by: Eduard Vintilă <eduard.vintila47@gmail.com>

mariasfiraiala reviewed (approved):
Thank you @eduardvintila for this fix.
For the past month I've been using the changes in this PR whenever I wanted to run `SQLite`, `redis` or`nginx` on `AArch64` and they work perfectly.

All good on my part, @michpappas your turn.

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

razvand commented:
@michpappas , could you please prioritize checking this PR?

michpappas commented on plat/drivers/virtio/virtio_pci.c:
```diff
@@ -273,10 +273,12 @@ static int vpci_legacy_pci_config_get(struct virtio_dev *vdev, __u16 offset,
 				VIRTIO_PCI_CONFIG_OFF + offset, buf, len,
 				type_len);
 	} else {
+		__u32 buf_len_bytes = len * type_len;
```
Careful here: As `len` is also `__u32` you risk an overflow, so it would be better if you added a check against that. 
One way to do that is to check for reasonable results after multiplication, eg:
```
	unsigned long a = 1 << 31;
	unsigned long b = 10;
	unsigned long x = a * b;

	if (x / a != b)
		return -1;
```

A better way is to use `__builtin_umull_overflow()`, that is supported by both GCC and clang, and it compiles into checking the processor's carry flag. That avoids division, that is generally expensive.
```
	unsigned long a = 1 << 31;
	unsigned long b = 10;
	unsigned long x;
	
	if (__builtin_umulll_overflow(a, b, &x))
		return -1;
```

Notice that these bugs can potentially lead into security vulnerabilites, and that was the case with the infamous 2002 vulnerability in OpenSSH: http://undeadly.org/cgi?action=article&sid=20060330071917

michpappas commented on plat/drivers/virtio/virtio_pci.c:
```diff
@@ -273,10 +273,12 @@ static int vpci_legacy_pci_config_get(struct virtio_dev *vdev, __u16 offset,
-				VIRTIO_PCI_CONFIG_OFF + offset,	buf, len);
-		if (rc != (int)len)
+				VIRTIO_PCI_CONFIG_OFF + offset,	buf, buf_len_bytes);
+		if (rc != (int)(buf_len_bytes))
```
The parenthesis around `buf_len_bytes` here are not needed. And while you're there, maybe wrap the check around `unlikely` to help with performance a bit.

michpappas reviewed (changes requested):
Hi @eduardvintila, except from the comments below, please fix the issues reported by checkpatch.pl. Other than these it looks good 👍🏼

eduardvintila commented:
Hey, @michpappas! Very good point on checking the multiplication result, sorry for missing that. I've added an overflow check using a compiler built-in function, just as you suggested. I chose `__builtin_umul_overflow()`, which works on `unsigned int` types. Since the ABIs for all of our target architectures (`arm`, `x86_64` and `riscv64` in future) specify that `int` types are 32-bits wide, that should be well suited for our `__u32` multiplication.

I've also fixed the coding style issues that you have mentioned.

michpappas reviewed (approved):
@eduardvintila thanks a lot for the update, it all looks good now! 👍🏼 

Approved-by: Michalis Pappas <mpappas@fastmail.fm>

gabrielmocanu commented:
@razvand  Can we merge it?
